src/m3ninx/index/segment/builder/multi_segments_multi_key_postings_list_iter.go

```diff
@@ -0,0 +1,219 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package builder
+
+import (
+	"bytes"
+
+	"github.com/m3db/m3/src/m3ninx/index"
+	"github.com/m3db/m3/src/m3ninx/index/segment"
+	"github.com/m3db/m3/src/m3ninx/postings"
+	"github.com/m3db/m3/src/m3ninx/postings/roaring"
+	xerrors "github.com/m3db/m3/src/x/errors"
+	bitmap "github.com/m3db/pilosa/roaring"
+)
+
+var _ segment.FieldsPostingsListIterator = &multiKeyPostingsListIterator{}
+
+type multiKeyPostingsListIterator struct {
+	err                    error
+	firstNext              bool
+	closeIters             []keyIterator
+	iters                  []keyIterator
+	currIters              []keyIterator
+	currReaders            []index.Reader
+	segments               []segment.Segment
+	currFieldPostingsList  postings.MutableList
+	currFieldPostingsLists []postings.List
+}
+
+func newMultiKeyPostingsListIterator() *multiKeyPostingsListIterator {
+	b := bitmap.NewBitmapWithDefaultPooling(defaultBitmapContainerPooling)
+	i := &multiKeyPostingsListIterator{
+		currFieldPostingsList: roaring.NewPostingsListFromBitmap(b),
+	}
+	i.reset()
+	return i
+}
+
+func (i *multiKeyPostingsListIterator) reset() {
+	i.firstNext = true
+	i.currFieldPostingsList.Reset()
+
+	for j := range i.closeIters {
+		i.closeIters[j] = nil
+	}
+	i.closeIters = i.closeIters[:0]
+
+	for j := range i.iters {
+		i.iters[j] = nil
+	}
+	i.iters = i.iters[:0]
+
+	for j := range i.currIters {
+		i.currIters[j] = nil
+	}
+	i.currIters = i.currIters[:0]
+
+	for j := range i.segments {
+		i.segments[j] = nil
+	}
+	i.segments = i.segments[:0]
+}
+
+func (i *multiKeyPostingsListIterator) add(iter keyIterator, segment segment.Segment) {
+	i.closeIters = append(i.closeIters, iter)
+	i.iters = append(i.iters, iter)
+	i.segments = append(i.segments, segment)
+	i.tryAddCurr(iter)
+}
+
+func (i *multiKeyPostingsListIterator) Next() bool {
+	if i.err != nil {
+		return false
+	}
+	if len(i.iters) == 0 {
+		return false
+	}
+
+	if i.firstNext {
+		i.firstNext = false
+		return true
+	}
+
+	for _, currIter := range i.currIters {
+		currNext := currIter.Next()
+		if currNext {
+			// Next has a value, forward other matching too
+			continue
+		}
+
+		// Remove iter
+		n := len(i.iters)
+		idx := -1
+		for j, iter := range i.iters {
+			if iter == currIter {
+				idx = j
+				break
+			}
+		}
+		i.iters[idx] = i.iters[n-1]
+		i.iters[n-1] = nil
+		i.iters = i.iters[:n-1]
+	}
+	if len(i.iters) == 0 {
+		return false
+	}
+
+	prevField := i.currIters[0].Current()
+
+	// Re-evaluate current value
+	i.currEvaluate()
+
+	// NB(bodu): Build the postings list for this field if the field has changed.
+	i.currFieldPostingsLists = i.currFieldPostingsLists[:0]
+	i.currReaders = i.currReaders[:0]
```

Comment:
Probably best to reset these after use so the readers and postings lists aren't held onto by the arrays:
```go
defer func() {
  for i := range i. currFieldPostingsLists {
    i. currFieldPostingsLists[i] = nil
  }
  i. currFieldPostingsLists = i. currFieldPostingsLists[:0]
  for i := range i.currReaders {
    i.currReaders[i] = nil
  }
  i.currReaders = i.currReaders[:0]
}()
```